libs/sdk-ui-dashboard/src/presentation/widget/insightMenu/DefaultDashboardInsightMenu/DashboardInsightMenu/index.tsx

```diff
@@ -1,15 +1,108 @@
-import { DashboardInsightMenuItemButton } from "./DashboardInsightMenuItemButton.js";
-import { DashboardInsightSubmenuContainer } from "./DashboardInsightSubmenuContainer.js";
 import { selectRenderMode, useDashboardSelector } from "../../../../../model/index.js";
 import { DashboardInsightMenuBubble } from "./DashboardInsightMenuBubble.js";
 import { DashboardInsightEditMenuBubble } from "./DashboardInsightEditMenuBubble.js";
 import { RenderMode } from "../../../../../types.js";
+import {
+    CustomUiMenuContentItemComponent,
+    CustomUiMenuHeaderComponent,
+    CustomUiMenuInteractiveItemComponent,
+    CustomUiMenuStaticItemComponent,
+    CustomUiMenuContentComponent,
+    IMenuData,
+    IMenuItemData,
+} from "./CustomUiMenuComponents.js";
+import { objRefToString, widgetRef } from "@gooddata/sdk-model";
+
+const convertToUiMenuItems = (
+    items: IInsightMenuItem[],
+    widget: IDashboardInsightMenuProps["widget"],
+): Array<IUiMenuItem<IMenuItemData>> => {
+    return items.map((item): IUiMenuItem<IMenuItemData> => {
+        if (item.type === "separator") {
+            return {
+                type: "static" as const,
+                id: item.itemId,
+                data: null,
+            };
+        }
+
+        if (item.type === "group") {
+            return {
+                type: "group" as const,
+                id: item.itemId,
+                data: null,
+                stringTitle: item.itemName,
+                subItems: item.items ? convertToUiMenuItems(item.items, widget) : [],
+            };
+        }
+        const baseFocusableItem = {
+            id: item.itemId,
+            stringTitle: item.itemName,
+            isDisabled: item.disabled,
+            data: {
+                icon: item.icon,
+                className: item.className,
+                tooltip: item.tooltip,
+                onClick: item.onClick,
+            },
+        };
+
+        if (isIInsightMenuSubmenu(item)) {
+            if (item.items?.length) {
+                return {
+                    ...baseFocusableItem,
+                    type: "interactive" as const,
+                    subItems: convertToUiMenuItems(item.items, widget),
+                    data: {
+                        ...baseFocusableItem.data,
+                        subMenu: true,
+                    },
+                };
+            }
+            if (item.SubmenuComponent) {
+                const WrappedSubmenuComponent: React.FC<{ onBack?: () => void; onClose?: () => void }> = ({
+                    onBack,
+                    onClose,
+                }) => {
+                    const SubmenuComponent = item.SubmenuComponent!;
+                    return (
+                        <SubmenuComponent
+                            widget={widget}
+                            onClose={onClose ?? (() => {})}
+                            onGoBack={onBack ?? (() => {})}
```

Comment:
Wouldn't undefined be better? Why the empty function must be provided?

Reply:
callbacks are mandatory on interface, which is used in many other components and already marked as alpha, not internal. So I did not want to change it